community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/CompatibilityFor3_0.scala

```diff
@@ -56,14 +56,27 @@ object helpersv3_0 {
   implicit def monitorFailure(t: Throwable)(implicit monitor: QueryExecutionMonitor, session: QuerySession): Unit = {
     monitor.endFailure(session, t)
   }
+}
 
+object typeConversionsFor3_0 {
   def asPublicType(value: Any): Any = value match {
-    case p: Point => wrapPoint(p)
-
+    case p: Point => asPublicPoint(p)
     case other => other
   }
 
-  private def wrapPoint(point: Point) = new graphdb.spatial.Point {
+  def asPrivateType(obj: Any): Any = obj match {
+    case map: Map[String, Any] => asPrivateMap(map)
+    case seq: Seq[Any] => seq.map(asPrivateType)
+    case arr: Array[Any] => arr.map(asPrivateType)
+    case point: graphdb.spatial.Point => asPrivatePoint(point)
+    case geometry: graphdb.spatial.Geometry => asPrivateGeometry(geometry)
+    case value => value
+  }
+
+  def asPrivateMap(incoming: Map[String, Any]): Map[String, Any] =
+    incoming.mapValues(asPrivateType)
```

Comment:
I think you want to use `immutableMapValues(incoming, asDeepScalaValue)` here, `mapValues`  returns a lazy view of the map